### ISSUE TITLE
Exception thrown by stylish-haskell plugin on format

### ISSUE THREAD

**Issue by williamareynolds** [haskell/haskell-language-server#2698]
### Your environment

Which OS do you use: macOS

Which LSP client (editor/plugin) do you use: Doom Emacs + LSP Mode/lsp-haskell

Describe your project (alternative: link to the project):
[Minimal Reproduction](https://github.com/williamareynolds/hls-stylish-plugin-test)

### Steps to reproduce

1. Install haskell-language-server version 1.6.1.0, GHC 8.10.7, and stack 2.7.3, implicit-hie. and the latest Doom Emacs with `lsp`, `lsp-haskell`, and `format +onsave`.
1. Configure lsp-haskell as follows in the doom config:

```elisp
;; Haskell
;; Process
(after! lsp-haskell
        (setq haskell-process-type 'stack)
        (setq lsp-haskell-brittany-on nil)
        (setq lsp-haskell-fourmolu-on nil)
        (setq lsp-haskell-ormolu-on nil)
        (setq lsp-haskell-floskell-on nil)
        (setq lsp-ui-doc-show-with-cursor t)
        (setq lsp-haskell-formatting-provider "stylish-haskell")
        (setq flycheck-hlintrc ".hlint.yaml"))
```

1. Build the project using `stack build`
1. Generate hie.yaml if necessary `gen-hie > hie.yaml`
1. Open app/Main.hs in a buffer, make a modification, and save. Everything should work fine.
1. Open other-app/Main.hs in a buffer, make a modification, and save. You will get an error: 

`Error: (doom-hook-error before-save-hook format-all-buffer--from-hook (error "Exception in plugin PluginId \"stylish-haskell\"while processing STextDocumentFormatting: BadDependency \"GetModSummary\""))`

### Expected behaviour
The stylish-haskell plugin should be able to format the file. Other formatters work correctly (at least, the Ormolu plugin does).

### Actual behaviour
Formatting does not occur and an error is written to the Emacs message buffer:

`Error: (doom-hook-error before-save-hook format-all-buffer--from-hook (error "Exception in plugin PluginId \"stylish-haskell\"while processing STextDocumentFormatting: BadDependency \"GetModSummary\""))`

### Include debug information

--probe-tools output:
```
haskell-language-server version: 1.6.1.0 (GHC: 8.10.7) (PATH: /Users/r258625/.ghcup/bin/haskell-language-server-wrapper-1.6.1.0) (GIT hash: f4022c5bb8530cd306c53b941878244bf27a5d41)
Tool versions found on the $PATH
cabal:          3.6.2.0
stack:          2.7.3
ghc:            8.10.7
```

I couldn't find much else in the logs that looked problematic. 

One thing I did find is that this issue only seems to occur (at least consistently) when two executable entrypoints have the same file name. If I change `other-app/Main.hs` to `other-app/OtherMain.hs`, rebuild, regenerate the hie.yaml, and restart the language server, the issue appears to be resolved. I think the issue is that having more than one file with the same name could be breaking the plugin. It also seems that this is only an issue with executable entry points. There are no issues with two files in the lib being called Lib.hs and Foo/Lib.hs for instance.

**Comment by Ailrun** [comment]
It looks like this is either macOS specific or Doom Emacs specific. I couldn't reproduce this with my vanilla Emacs setting.

**Comment by Ailrun** [comment]
@williamareynolds @drsooch could anyone try the same with Brittany plugin? I think
https://github.com/haskell/haskell-language-server/blob/4b8f9d53bed3e775ccd120d9a2471ae374a88db3/plugins/hls-brittany-plugin/src/Ide/Plugin/Brittany.hs#L67
may cause the same issue

**Comment by Ailrun** [comment]
@michaelpj
> nothing the client does should make us crash!

Nothing should, but I'm just saying that there are those two possibility, as I couldn't reproduce it in Emacs/Linux with the same ghcup/cabal/stack/ghc version.

**Comment by williamareynolds** [comment]
@Ailrun I can confirm that this issue does occur with Brittany as well with all else being equal.

**Comment by williamareynolds** [comment]
> It looks like this is either macOS specific or Doom Emacs specific. I couldn't reproduce this with my vanilla Emacs setting.

I have confirmed that the issue occurs on vanilla emacs in macOS.

**Comment by Ailrun** [comment]
Thank you for the check! Just to be sure, which method did you use to reproduce it in your vanilla Emacs? Just by invoking `lsp-format-buffer`?

**Comment by williamareynolds** [comment]
> Thank you for the check! Just to be sure, which method did you use to reproduce it in your vanilla Emacs? Just by invoking `lsp-format-buffer`?

That's correct. In my initial attempt I tried with `lsp-format-buffer` and was seeing the same issue with the format-on-save hook, which was just delegating to `lsp-format-buffer`.

**Comment by michaelpj** [comment]
Note to self: we don't *crash*, this is a properly caught and reported exception.

Poking around a little, it looks to me like this happens if we can't compute `GetModSummary` at all. Now I'm guessing there might be something in the HLS log - can you have a look?

(Separately: @pepeiborra do you think it would be feasible to include some kind of summary about why the rule failed when we record a `Failed` value? Then we could also report it in the `BadDependency` error.)

**Comment by williamareynolds** [comment]
I actually managed to catch it happening, but it was a bit uncontrolled in another project. Pardon the redactions:

```
	Set files of interest to: fromList [(NormalizedFilePath "/Users/<redacted>/code/<redacted>/src/Server/Handler/<redacted>.hs",OnDisk),(NormalizedFilePath "/Users/<redacted>/code/<redacted>/ServiceCheck.hs",OnDisk),(NormalizedFilePath "/Users/<redacted>/code/<redacted>/Publish.hs",Modified {firstOpen = False}),(NormalizedFilePath "/Users/<redacted>/code/<redacted>/Main.hs",OnDisk)]
2022-02-10 13:36:25.044528 [ThreadId 69184] DEBUG hls:	Finishing build session(exception: AsyncCancelled)
2022-02-10 13:36:25.044581 [ThreadId 25] DEBUG hls:	Restarting build session due to /Users/<redacted>/code/<redacted>/Main.hs (modified) with queue [] for keys [GetModificationTime; /Users/<redacted>/code/<redacted>/Main.hs,IsFileOfInterest; /Users/<redacted>/code/<redacted>/Main.hs,GetModificationTime; /Users/<redacted>/.cache/ghcide/main-b906ff1b70360176482b3b0bd4c5f4ee3df714af/Server/Handler/<redacted>.hi] (aborting the previous one took 0.00s)
2022-02-10 13:36:25.044728 [ThreadId 25] DEBUG hls:	Saved text document: file:///Users/<redacted>/code/<redacted>/Main.hs
2022-02-10 13:36:25.045009 [ThreadId 69382] INFO hls:	Could not identify reverse dependencies for NormalizedFilePath "/Users/<redacted>/code/<redacted>/Main.hs"
2022-02-10 13:36:25.045096 [ThreadId 69382] DEBUG hls:	finish: ParentTC (took 0.00s)
```

**Comment by drsooch** [comment]
> @drsooch, did you reproduce it in macOS, or in Linux?

Linux. I took a look at the logs last night didn't see anything out of the ordinary. I'll try with another formatter tonight. 

Judging by the code I doubt this would be why, but I notice it more often with CPP enabled files. That might also just be recency bias as I've been stuck in those types of files recently.

**Comment by drsooch** [comment]
I updated my hie.yaml as well seemed to resolve my issue. Realized it was out of date in between hopping branches. Issue appears to be gone.

**Comment by fosskers** [comment]
This still appears to be happening.
